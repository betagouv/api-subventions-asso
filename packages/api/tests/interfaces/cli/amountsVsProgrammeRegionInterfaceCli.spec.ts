import { ObjectId } from "mongodb";
import amountsVsProgrammeRegionPort from "../../../src/dataProviders/db/dataViz/amountVSProgrammeRegion/amountsVsProgrammeRegion.port";
import { PAYMENT_FLAT_DBO } from "../../../src/dataProviders/db/paymentFlat/__fixtures__/paymentFlatDbo.fixture";
import paymentFlatPort from "../../../src/dataProviders/db/paymentFlat/paymentFlat.port";
import AmountsVsProgrammeRegionCli from "../../../src/interfaces/cli/AmountsVsProgrammeRegion.cli";

const MOCK_DOCUMENTS = [
    { ...PAYMENT_FLAT_DBO, provider: "Chorus", uniqueId: "1" },
    { ...PAYMENT_FLAT_DBO, montant: 7000, provider: "Chorus", uniqueId: "2" },
    { ...PAYMENT_FLAT_DBO, montant: 100_000, numeroProgramme: "1234", provider: "Chorus", uniqueId: "3" },
    {
        ...PAYMENT_FLAT_DBO,
        montant: 34_000,
        attachementComptable: "HNOR",
        exerciceBudgetaire: 2021,
        provider: "Chorus",
        uniqueId: "4",
    },
];

const insertData = async () => {
    const bulkWriteArray = MOCK_DOCUMENTS.map(dbo => {
        const { _id, ...DboWithoutId } = dbo;
        return {
            updateOne: {
                filter: { uniqueId: DboWithoutId.uniqueId },
                update: { $set: DboWithoutId },
                upsert: true,
            },
        };
    });

    await paymentFlatPort.upsertMany(bulkWriteArray);
};

describe("AmountsVsProgrammeRegionCli", () => {
    beforeEach(async () => {
        await amountsVsProgrammeRegionPort.deleteAll();
        await insertData();
    });

    afterEach(() => {
        jest.restoreAllMocks();
    });

    let cli = new AmountsVsProgrammeRegionCli();

    describe("init()", () => {
        it("should persist amounts vs programme region collection", async () => {
            await cli.init();
            //@ts-expect-error protected method
            const amountsVsProgrammeRegion = (await amountsVsProgrammeRegionPort.collection.find({}).toArray())
                .map(amountsVsProgrammeRegion => ({
                    ...amountsVsProgrammeRegion,
                    _id: expect.any(ObjectId),
                }))
                .sort((a, b) => {
                    if (a.regionAttachementComptable < b.regionAttachementComptable) return -1;
                    if (a.regionAttachementComptable > b.regionAttachementComptable) return 1;
                    if (a.programme < b.programme) return -1;
                    if (a.programme > b.programme) return 1;

                    return a.exerciceBudgetaire - b.exerciceBudgetaire;
                });

            expect(amountsVsProgrammeRegion).toMatchSnapshot("Snapshot init");
        });
    });

    describe("resyncExercice()", () => {
        it("should persist amounts vs programme region collection for the given exercice", async () => {
            const exercice = 2023;
            await cli.resyncExercice(exercice);
            //@ts-expect-error protected method
            const amountsVsProgrammeRegion = (await amountsVsProgrammeRegionPort.collection.find({}).toArray())
                .map(amountsVsProgrammeRegion => ({
                    ...amountsVsProgrammeRegion,
                    _id: expect.any(ObjectId),
                }))
                .sort((a, b) => {
                    if (a.regionAttachementComptable < b.regionAttachementComptable) return -1;
                    if (a.regionAttachementComptable > b.regionAttachementComptable) return 1;
                    if (a.programme < b.programme) return -1;
                    if (a.programme > b.programme) return 1;

                    return a.exerciceBudgetaire - b.exerciceBudgetaire;
                });

            expect(amountsVsProgrammeRegion).toMatchSnapshot("Snapshot resyncExercice");
        });
    });
});
