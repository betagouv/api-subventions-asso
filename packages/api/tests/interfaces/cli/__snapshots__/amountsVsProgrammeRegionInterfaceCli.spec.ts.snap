// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AmountsVsProgrammeRegionCli init() should persist amounts vs programme region collection: Snapshot init 1`] = `
Array [
  Object {
    "_id": Any<ObjectId>,
    "amount": 8000,
    "exerciceBudgetaire": 2023,
    "mission": "Mission Exemple",
    "programme": "1 - Programme Exemple",
    "regionAttachementComptable": "Bretagne",
  },
  Object {
    "_id": Any<ObjectId>,
    "amount": 100000,
    "exerciceBudgetaire": 2023,
    "mission": "Mission Exemple",
    "programme": "1234 - Programme Exemple",
    "regionAttachementComptable": "Bretagne",
  },
  Object {
    "_id": Any<ObjectId>,
    "amount": 34000,
    "exerciceBudgetaire": 2021,
    "mission": "Mission Exemple",
    "programme": "1 - Programme Exemple",
    "regionAttachementComptable": "Normandie",
  },
]
`;

exports[`AmountsVsProgrammeRegionCli resyncExercice() should persist amounts vs programme region collection for the given exercice: Snapshot resyncExercice 1`] = `
Array [
  Object {
    "_id": Any<ObjectId>,
    "amount": 8000,
    "exerciceBudgetaire": 2023,
    "mission": "Mission Exemple",
    "programme": "1 - Programme Exemple",
    "regionAttachementComptable": "Bretagne",
  },
  Object {
    "_id": Any<ObjectId>,
    "amount": 100000,
    "exerciceBudgetaire": 2023,
    "mission": "Mission Exemple",
    "programme": "1234 - Programme Exemple",
    "regionAttachementComptable": "Bretagne",
  },
]
`;
